materialize-boilerplate/validate.go

```diff
@@ -203,26 +185,87 @@ func (v Validator) validateMatchesExistingBinding(
 					),
 				}
 			}
+		} else if ambiguousFields := v.ambiguousFields(p, boundCollection.Projections); len(ambiguousFields) > 0 {
+			// Projections that would result in ambiguous materialized fields are forbidden if a
+			// different ambiguous projection has already been selected, or optional if none have
+			// yet been selected.
+			if existing != nil && slices.Contains(existing.FieldSelection.AllFields(), p.Field) {
+				// This field has already been selected as the ambiguous field to materialize.
+				c = &pm.Response_Validated_Constraint{
+					Type:   pm.Response_Validated_Constraint_LOCATION_RECOMMENDED,
+					Reason: "This location is part of the current materialization",
+				}
+			} else if existing != nil && v.ambiguousFieldIsSelected(p, existing.FieldSelection.AllFields()) {
+				// A different field has been selected, so this one can't be.
+				c = &pm.Response_Validated_Constraint{
+					Type: pm.Response_Validated_Constraint_FIELD_FORBIDDEN,
+					Reason: fmt.Sprintf(
+						"Flow collection field '%s' is ambiguous with fields already being materialized as '%s' in the destination. Consider using an alternate, unambiguous projection of this field to allow it to be materialized",
```

Comment:
:+1: all this was no-doubt tedious to implement, but will be really helpful for users trying to understand what's happening and why.